src/modules/robot/Robot.cpp

```diff
@@ -194,58 +205,53 @@ void Robot::on_config_reload(void *argument)
     this->max_speeds[Y_AXIS]  = THEKERNEL->config->value(y_axis_max_speed_checksum    )->by_default(60000.0F)->as_number() / 60.0F;
     this->max_speeds[Z_AXIS]  = THEKERNEL->config->value(z_axis_max_speed_checksum    )->by_default(  300.0F)->as_number() / 60.0F;
 
-    Pin alpha_step_pin;
-    Pin alpha_dir_pin;
-    Pin alpha_en_pin;
-    Pin beta_step_pin;
-    Pin beta_dir_pin;
-    Pin beta_en_pin;
-    Pin gamma_step_pin;
-    Pin gamma_dir_pin;
-    Pin gamma_en_pin;
-
-    alpha_step_pin.from_string( THEKERNEL->config->value(alpha_step_pin_checksum )->by_default("2.0"  )->as_string())->as_output();
-    alpha_dir_pin.from_string(  THEKERNEL->config->value(alpha_dir_pin_checksum  )->by_default("0.5"  )->as_string())->as_output();
-    alpha_en_pin.from_string(   THEKERNEL->config->value(alpha_en_pin_checksum   )->by_default("0.4"  )->as_string())->as_output();
-    beta_step_pin.from_string(  THEKERNEL->config->value(beta_step_pin_checksum  )->by_default("2.1"  )->as_string())->as_output();
-    beta_dir_pin.from_string(   THEKERNEL->config->value(beta_dir_pin_checksum   )->by_default("0.11" )->as_string())->as_output();
-    beta_en_pin.from_string(    THEKERNEL->config->value(beta_en_pin_checksum    )->by_default("0.10" )->as_string())->as_output();
-    gamma_step_pin.from_string( THEKERNEL->config->value(gamma_step_pin_checksum )->by_default("2.2"  )->as_string())->as_output();
-    gamma_dir_pin.from_string(  THEKERNEL->config->value(gamma_dir_pin_checksum  )->by_default("0.20" )->as_string())->as_output();
-    gamma_en_pin.from_string(   THEKERNEL->config->value(gamma_en_pin_checksum   )->by_default("0.19" )->as_string())->as_output();
-
-    float steps_per_mm[3] = {
-        THEKERNEL->config->value(alpha_steps_per_mm_checksum)->by_default(  80.0F)->as_number(),
-        THEKERNEL->config->value(beta_steps_per_mm_checksum )->by_default(  80.0F)->as_number(),
-        THEKERNEL->config->value(gamma_steps_per_mm_checksum)->by_default(2560.0F)->as_number(),
-    };
-
     // TODO: delete or detect old steppermotors
+#if 0 //TODO: Add unregistering to StepperMotor/StepTicker so we can delete old motors 
+    for (auto a : actuators)
+        delete a;
+#endif
+    actuators.clear();
+
     // Make our 3 StepperMotors
-    this->alpha_stepper_motor  = new StepperMotor(alpha_step_pin, alpha_dir_pin, alpha_en_pin);
-    this->beta_stepper_motor   = new StepperMotor(beta_step_pin,  beta_dir_pin,  beta_en_pin );
-    this->gamma_stepper_motor  = new StepperMotor(gamma_step_pin, gamma_dir_pin, gamma_en_pin);
+    struct
+    {
+        uint16_t pin_checksums[3];
+        uint16_t per_mm_checksum;
+        uint16_t per_sec_checksum;
+        char const *pin_defaults[3];
+    } const p[] =
+    {
+        { ACTUATOR_CHECKSUMS("alpha"),   {"2.0", "0.5",  "0.4" } },
```

Comment:
Without the defaults, I was able to flatten the table. Also compiled out the extra rows when k_max_actuators == 3.Still not ideal. I would suggest smoothie2 use the dot style configuation for actuators, which would make this table unnecessary and the firmware more compact.
      From: Jim Morris <notifications@github.com>
 To: Smoothieware/Smoothieware <Smoothieware@noreply.github.com> 
Cc: trayracing <daqray@yahoo.com> 
 Sent: Tuesday, November 10, 2015 3:37 AM
 Subject: Re: [Smoothieware] Feature/nmotors (#759)
   
In src/modules/robot/Robot.cpp:>      // Make our 3 StepperMotors
> -    this->alpha_stepper_motor  = new StepperMotor(alpha_step_pin, alpha_dir_pin, alpha_en_pin);
> -    this->beta_stepper_motor   = new StepperMotor(beta_step_pin,  beta_dir_pin,  beta_en_pin );
> -    this->gamma_stepper_motor  = new StepperMotor(gamma_step_pin, gamma_dir_pin, gamma_en_pin);
> +    struct
> +    {
> +        uint16_t pin_checksums[3];
> +        uint16_t per_mm_checksum;
> +        uint16_t per_sec_checksum;
> +        char const *pin_defaults[3];
> +    } const p[] =
> +    {
> +        { ACTUATOR_CHECKSUMS("alpha"),   {"2.0", "0.5",  "0.4" } },
I'm not too fond of this construct, however I can live with it. But I suggest removing the default pins, and just set them to "nc" if undefined. It should be ok as every config file has always had them specifically defined. This will save some flash and make this construct a bit more tolerable. - Thanks—
Reply to this email directly or view it on GitHub.